RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT.cc

```diff
@@ -0,0 +1,778 @@
+  }
+}
+
+
+
+
+
+vector< TransientVertex >
+DAClusterizerInZT::vertices(const vector<reco::TransientTrack> & tracks, const int verbosity) 
+const
+{
+ 
+  vector<track_t> tks=fill(tracks);
+  unsigned int nt=tks.size();
+  double rho0=0.0;  // start with no outlier rejection
+
+  vector< TransientVertex > clusters;
+  if (tks.empty()) return clusters;
+
+  vector<vertex_t> y; // the vertex prototypes
+
+  // initialize:single vertex at infinite temperature
+  vertex_t vstart;
+  vstart.z=0.;
+  vstart.t=0.;
+  vstart.pk=1.;
+  y.push_back(vstart);
+  int niter=0;      // number of iterations
+  
+
+  // estimate first critical temperature
+  double beta=beta0(betamax_, tks, y);
+  niter=0; while((update(beta, tks,y)>1.e-6)  && (niter++ < maxIterations_)){ }
+
+  // annealing loop, stop when T<Tmin  (i.e. beta>1/Tmin)
+  while(beta<betamax_){ 
+
+    if(useTc_){
+      update(beta, tks,y);
+      while(merge(y,beta)){update(beta, tks,y);}
+      split(beta, tks,y, 1.);
+      beta=beta/coolingFactor_;
+    }else{
+      beta=beta/coolingFactor_;
+      splitAll(y);
+    }
+
+
+   // make sure we are not too far from equilibrium before cooling further
+   niter=0; while((update(beta, tks,y)>1.e-6)  && (niter++ < maxIterations_)){ }
+
+  }
+
+  if(useTc_){
+    // last round of splitting, make sure no critical clusters are left
+    update(beta, tks,y);
+    while(merge(y,beta)){update(beta, tks,y);}
+    unsigned int ntry=0;
+    while( split(beta, tks,y,1.) && (ntry++<10) ){
+      niter=0; 
+      while((update(beta, tks,y)>1.e-6)  && (niter++ < maxIterations_)){}
+      merge(y,beta);
+      update(beta, tks,y);
+    }
+  }else{
+    // merge collapsed clusters 
+    while(merge(y,beta)){update(beta, tks,y);}  
+    if(verbose_ ){ cout << "dump after 1st merging " << endl;  dump(beta,y,tks,2);}
+  }
+  
+
+
+
+  // switch on outlier rejection
+  rho0=1./nt; for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); k++){ k->pk =1.; }  // democratic
+  niter=0; while((update(beta, tks,y,rho0) > 1.e-8)  && (niter++ < maxIterations_)){  }
+  if(verbose_  ){ cout << "rho0=" << rho0 <<   " niter=" << niter <<  endl; dump(beta,y,tks,2);}
+
+  
+  // merge again  (some cluster split by outliers collapse here)
+  while(merge(y,tks.size())){}  
+  if(verbose_  ){ cout << "dump after 2nd merging " << endl;  dump(beta,y,tks,2);}
+
+
+  // continue from freeze-out to Tstop (=1) without splitting, eliminate insignificant vertices
+  while(beta<=betastop_){
+    while(purge(y,tks,rho0, beta)){
+      niter=0; while((update(beta, tks, y, rho0) > 1.e-6)  && (niter++ < maxIterations_)){  }
+    } 
+    beta/=coolingFactor_;
+    niter=0; while((update(beta, tks, y, rho0) > 1.e-6)  && (niter++ < maxIterations_)){  }
+  }
+
+
+//   // new, one last round of cleaning at T=Tstop
+//   while(purge(y,tks,rho0, beta)){
+//     niter=0; while((update(beta, tks,y,rho0) > 1.e-6)  && (niter++ < maxIterations_)){  }
+//   } 
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant